### PR TITLE
ENH: Roll moment calculation due to fin cant angle

### PR DESCRIPTION
**Pull request by MateusStano**
## Pull request type

Please check the type of change your PR introduces:

- [x] Code base additions (bugfix, features)
- [ ] Code maintenance (refactoring, formatting, renaming, tests)
- [ ] ReadMe, Docs and GitHub maintenance
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements, depending on the type of PR:

- ReadMe, Docs and GitHub maintenance:

  - [ ] Spelling has been verified
  - [ ] Code docs are working correctly 

- Code base maintenance (refactoring, formatting, renaming):

  - [ ] Docs have been reviewed and added / updated if needed
  - [ ] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [ ] All tests (`pytest --runslow`) have passed locally

- Code base additions (for bug fixes / features):

  - [ ] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [x] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

Currently RocketPy doesn't calculate roll moment due to fin cant angle

## What is the new behavior?

This addition allows a cant angle input when adding a fin group to a rocket. Then, roll moment is calculated accordingly

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

I believe the simulations results with the cant angle are good but I am not 100% sure. If anyone could double check that would be great!

### CODE REVIEW

**Comment by giovaniceotto** [comment]
Awesome work @MateusStano! I'll review it ASAP.

File: rocketpy/Rocket.py [hunk 1]
```diff
@@ -593,8 +615,18 @@ def addFins(
...
 
+            # Parameters for Roll Moment
+            if delta != 0:
+                trapezoidalConstant = ((Yr) / 2) * (radius ** 2) * s
+                trapezoidalConstant += ((Cr + 2 * Ct) / 3) * radius * (s ** 2)
+                trapezoidalConstant += ((Cr + 3 * Ct) / 12) * (s ** 3)
+
+                clfDelta = n * (Ymac + radius) * clalpha / d
+                cldOmega = n * clalpha * np.cos(delta) * trapezoidalConstant / (Af * d)
+                rollParameters = [clfDelta, cldOmega, delta]
```
Comment (giovaniceotto) [inline rocketpy/Rocket.py:626]:
```suggestion
                cldOmega = n * clalpha * np.cos(cantAngleRad) * trapezoidalConstant / (Af * d)
                rollParameters = [clfDelta, cldOmega, cantAngleRad]
```

File: rocketpy/Rocket.py [hunk 2]
```diff
@@ -652,8 +684,20 @@ def cnalfa1(cn):
...
+            if delta != 0:
+                trapezoidalConstant = ((Yr) / 2) * (radius ** 2) * s
+                trapezoidalConstant += ((Cr + 2 * Ct) / 3) * radius * (s ** 2)
+                trapezoidalConstant += ((Cr + 3 * Ct) / 12) * (s ** 3)
+
+                clfDelta = n * (Ymac + radius) * clalpha / d
+
+                cldOmega = n * clalpha * np.cos(delta) * trapezoidalConstant / (Af * d)
+
+                rollParameters = [clfDelta, cldOmega, delta]
```
Comment (giovaniceotto) [inline rocketpy/Rocket.py:697]: This is a repeated code, which can be hard to maintain (we may change its first appearance above and forget to change this one).

Is there any way for us to remove this part of the code from the "airfoil if statement" and thus make it appear only once? Perhaps we can treat all fins as airfoils and have a constant 2pi as the cl_alpha for the flat plate airfoil (which would mean setting a default 2*np.pi as the airfoil entry, something like that, or even change the name of the "airfoil" entry to "cl_alpha_curve")

Anyways, just suggesting ideas to improve code readability and mantainability.

File: rocketpy/Flight.py [hunk 3]
```diff
@@ -1339,6 +1339,13 @@ def uDot(self, t, u, postProcessing=False):
                     # Add to total moment
                     M1 -= (compCp + a) * compLiftYB
                     M2 += (compCp + a) * compLiftXB
+            # Calculates Roll Moment
+            if aerodynamicSurface[-1] == "Fins":
+                Clfdelta, Cldomega, delta = aerodynamicSurface[2]
+                if delta != 0:
+                    Clf = Clfdelta * delta
```
Comment (giovaniceotto) [inline rocketpy/Flight.py:1346]:
```suggestion
                Clfdelta, Cldomega, cantAngleRad = aerodynamicSurface[2]
                if cantAngleRad != 0:
                    Clf = Clfdelta * cantAngleRad
```

**Review by giovaniceotto — CHANGES_REQUESTED** [review]
@MateusStano, great contribution overall. This is an important feature that RocketPy was missing, and now we are close to having it! Great holiday season gift for the community.

I have added a couple of comments and code change suggestions which focus on improving code readability and maintainability. I ask you to look at them and decide what would be best for the future.

Out of curiosity, does this PR have anything to do with #109?

File: rocketpy/Rocket.py [hunk 4]
```diff
@@ -593,8 +615,18 @@ def addFins(
                 lambda x: clalpha * x, "Alpha (rad)", "Cl", interpolation="linear"
             )
 
+            # Parameters for Roll Moment
```
[the remark this answers is not in the document]
Comment (MateusStano) [reply rocketpy/Rocket.py]: I added a link to the documentation we made on this comment. Everything I used should be on there.

File: rocketpy/Rocket.py (hunk 2 above)
Comment (MateusStano) [reply rocketpy/Rocket.py]: I partially fixed this but this comment actually got me looking into the airfoil calculations again. I should make a PR soon changing this part of the code

**Comment by MateusStano** [comment]
> Out of curiosity, does this PR have anything to do with #109?

Yes! I actually started adding the roll before #109 but then I noticed some things were wrong in the airfoil calculations and got into that. I reviewed some stuff about airfoils now and I think it's not as hard of a fix as I thought it was. I will review that PR soon

**Review by giovaniceotto — APPROVED** [review]
Thanks for the modifications @MateusStano. Looks great.

Only two comments:

 - Was the file `getting_startedTeste.ipynb` committed as a mistake? I am not sure it should be part of this PR.
 - Is there any chance that the GDrive link added may be broken some day? I wonder if we should create a folder in this repo for similar documentation and store a copy of such files there, so that we can guarantee that as long as the code exists, they exist. Example: `docs/technical/aerodynamics/roll.pdf`.

Once you sort out these two comments, from my part, it is a GO to merging this PR.

**Comment by giovaniceotto** [comment]
@MateusStano, it seems that Travis CI tests are failing due to linting problems. Don't forget to run [black](https://github.com/psf/black) inside the rocketpy folder and commit the format changes to the files you have modified.

**Comment by MateusStano** [comment]
> * Was the file `getting_startedTeste.ipynb` committed as a mistake? I am not sure it should be part of this PR.

It was a mistake. I deleted it now

> * I wonder if we should create a folder in this repo for similar documentation

I created the folder as you suggested. The only thing is that the link at the code's documentation will work when the PR is merged.

> Don't forget to run [black](https://github.com/psf/black) inside the rocketpy folder and commit the format changes to the files you have modified.

I've been running black and it says all is fine:
image

What I've been having problems with is when I use pytest --runslow:
image

Don't know if it has anything to do with it though.

**Comment by giovaniceotto** [comment]
Perfect @MateusStano. I will look into the slow tests problems. Feel free to merge this PR into develop!

**Comment by Gui-FernandesBR** [comment]
@giovaniceotto @MateusStano , these calculations are going to be performed automatically when running rocketpy? Or the user needs to specify whether he wants to evaluate roll moment or not ?

**Comment by MateusStano** [comment]
@Gui-FernandesBR the roll moment is only calculated when a cant angle is added to the fins otherwise it will just be zero. So yeah, the user can indirectly specify with he wants roll to be calculated or not

**Comment by Gui-FernandesBR** [comment]
Ok tks, nice to hear that.

I remeber we had some problems in the past due to time consumption when running roll mometn calculation. I didn't test if it's happeninig with #116, but just in case, if anybody starts to complain about it, we could deep dive here

**Pull request merged**